source/common/network/buffered_io_socket_handle_impl.cc

```diff
@@ -0,0 +1,266 @@
+                               IoSocketError::deleteIoError)};
+  }
+  uint64_t total_bytes_to_write = buffer.length();
+  writable_peer_->getWriteBuffer()->move(buffer);
+  writable_peer_->maybeSetNewData();
+  ENVOY_LOG(trace, "socket {} writev {} bytes", static_cast<void*>(this), total_bytes_to_write);
+  return {total_bytes_to_write, Api::IoErrorPtr(nullptr, [](Api::IoError*) {})};
+}
+
+Api::IoCallUint64Result BufferedIoSocketHandleImpl::sendmsg(const Buffer::RawSlice*, uint64_t, int,
+                                                            const Address::Ip*,
+                                                            const Address::Instance&) {
+  return IoSocketError::ioResultSocketInvalidAddress();
+}
+
+Api::IoCallUint64Result BufferedIoSocketHandleImpl::recvmsg(Buffer::RawSlice*, const uint64_t,
+                                                            uint32_t, RecvMsgOutput&) {
+  return IoSocketError::ioResultSocketInvalidAddress();
+}
+
+Api::IoCallUint64Result BufferedIoSocketHandleImpl::recvmmsg(RawSliceArrays&, uint32_t,
+                                                             RecvMsgOutput&) {
+  return IoSocketError::ioResultSocketInvalidAddress();
+}
+
+Api::IoCallUint64Result BufferedIoSocketHandleImpl::recv(void* buffer, size_t length, int flags) {
+  if (!isOpen()) {
+    return {0,
+            Api::IoErrorPtr(new IoSocketError(SOCKET_ERROR_INVAL), IoSocketError::deleteIoError)};
+  }
+  // No data and the writer closed.
+  if (owned_buffer_.length() == 0) {
+    if (read_end_stream_) {
+      return {0, Api::IoErrorPtr(nullptr, [](Api::IoError*) {})};
+    } else {
+      return {0, Api::IoErrorPtr(IoSocketError::getIoSocketEagainInstance(),
+                                 IoSocketError::deleteIoError)};
+    }
+  } else {
+    auto max_bytes_to_read = std::min(owned_buffer_.length(), length);
+    owned_buffer_.copyOut(0, max_bytes_to_read, buffer);
+    if (!(flags & MSG_PEEK)) {
+      owned_buffer_.drain(max_bytes_to_read);
+    }
+    return {max_bytes_to_read, Api::IoErrorPtr(nullptr, [](Api::IoError*) {})};
+  }
+}
+
+bool BufferedIoSocketHandleImpl::supportsMmsg() const { return false; }
+
+bool BufferedIoSocketHandleImpl::supportsUdpGro() const { return false; }
+
+Api::SysCallIntResult makeInvalidSyscall() {
+  return Api::SysCallIntResult{-1, SOCKET_ERROR_NOT_SUP /*SOCKET_ERROR_NOT_SUP*/};
+}
+
+Api::SysCallIntResult BufferedIoSocketHandleImpl::bind(Address::InstanceConstSharedPtr) {
+  return makeInvalidSyscall();
+}
+
+Api::SysCallIntResult BufferedIoSocketHandleImpl::listen(int) { return makeInvalidSyscall(); }
+
+IoHandlePtr BufferedIoSocketHandleImpl::accept(struct sockaddr*, socklen_t*) {
+  NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+}
+
+Api::SysCallIntResult BufferedIoSocketHandleImpl::connect(Address::InstanceConstSharedPtr) {
+  // Buffered Io handle should always be considered as connected. Use write to determine if peer is
```

Comment:
Technically you can use both read and write to determine if peer is closed.

Reply:
TIL

Reply:
comment updated